pkg/core/native/management.go

```diff
@@ -334,19 +365,54 @@ func (m *Management) OnPersist(ic *interop.Context) error {
 		if err := native.Initialize(ic); err != nil {
 			return fmt.Errorf("initializing %s native contract: %w", md.Name, err)
 		}
+		m.mtx.Lock()
+		m.contracts[md.Hash] = cs
+		m.mtx.Unlock()
 	}
 
 	return nil
 }
 
 // PostPersist implements Contract interface.
-func (m *Management) PostPersist(_ *interop.Context) error {
+func (m *Management) PostPersist(ic *interop.Context) error {
+	m.mtx.Lock()
+	for h, cs := range m.contracts {
+		if cs != nil {
+			continue
+		}
+		newCs, err := m.getContractFromDAO(ic.DAO, h)
+		if err != nil {
+			// Contract was destroyed.
+			delete(m.contracts, h)
+			continue
+		}
+		m.contracts[h] = newCs
+	}
+	m.mtx.Unlock()
 	return nil
 }
 
 // Initialize implements Contract interface.
-func (m *Management) Initialize(_ *interop.Context) error {
-	return nil
+func (m *Management) Initialize(ic *interop.Context) error {
+	var initErr error
+	ic.DAO.Seek(m.ContractID, []byte{prefixContract}, func(_, v []byte) {
```

Comment:
If I'm not mistaken, going to DAO here almost does not make sense. Take a look at the order of native contracts initialisation: https://github.com/nspcc-dev/neo-go/blob/master/pkg/core/native/management.go#L321. They are ranged, and the first in the list is Management contract (see [here](https://github.com/nspcc-dev/neo-go/blob/master/pkg/core/native/contract.go#L60)). So Managment contract initialisation occurs when there's no other contracts in DAO yet => only management contract will be cached during `(m *Management) Initialize` call.

Reply:
Fixed. `Dump/Restore` is not exactly the test we want here, added a separate test for starting from height. Related #1577 .